### PR TITLE
Remove vestigial diverging !-coercion

### LINKED ISSUE #50350: we still sometimes permit coercion into `!`
While looking into the precise behavior around `!` for https://github.com/rust-lang/rust/pull/47291, we realized that we still sometimes permit coercion *into* `!`, in the `coerce_lub` code:

https://github.com/rust-lang/rust/blob/25749ad66de71ffb46e6b04fb0cccf1ae4fc5ed5/src/librustc_typeck/check/mod.rs#L3541-L3544

@arielb1 tried to remove that earlier https://github.com/rust-lang/rust/pull/45880 but we missed this case.

### PR DESCRIPTION
**Pull request by varkor**
Such conversions are no longer permitted. Fixes #50350.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc_typeck/check/_match.rs [hunk 1]
```diff
@@ -690,7 +690,7 @@ https://doc.rust-lang.org/reference/types.html#trait-objects");
                     arm_span: arm.body.span,
                     source: match_src
                 });
-                coercion.coerce(self, &cause, &arm.body, arm_ty, self.diverges.get());
+                coercion.coerce(self, &cause, &arm.body, arm_ty);
```
Comment (nikomatsakis) [inline src/librustc_typeck/check/_match.rs]: 👍 -- the fact that this parameter was still needed should have been a warning sign. sigh.

File: src/test/compile-fail/coerce-to-bang.rs [hunk 2]
```diff
@@ -56,9 +56,8 @@ fn call_foo_f() {
 }
 
 fn array_a() {
-    // Accepted: return is coerced to `!` just fine, and then `22` can be
-    // because we already diverged.
-    let x: [!; 2] = [return, 22];
+    // Accepted: return is coerced to `!` just fine, but `22` cannot be.
```
Comment (nikomatsakis) [inline src/test/compile-fail/coerce-to-bang.rs]: Nit: This is no longer "accepted"

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
One nit.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
cc @rust-lang/infra -- this could use a "check-only" crater run.

**Comment by aidanhs** [comment]
Crater run started

**Comment by kennytm** [comment]
We're still uploading to http://cargobomb-reports.s3.amazonaws.com/pr-50351/index.html, but there's only a single regression with `yamlette-0.0.4` (not sure if legit or spurious yet).

**Comment by aidanhs** [comment]
Upload complete at http://cargobomb-reports.s3.amazonaws.com/pr-50351/index.html

(sorry for the delay, uploading hit a bug in crater that I had to debug and fix)

**Comment by aidanhs** [comment]
Reuploaded, the report generation was buggy too (so both logs were identical). Regarding the timeout - seems like normal variance, looking at the two logs it appears the one that passed squeaked by with just a few seconds under 300.

**Pull request merged**